### PR TITLE
Configurable service logger names / Simpler construction of Server

### PR DESCRIPTION
**Pull request by trustin**
Motivation:

- ServiceInvocationContext.logger() returns a logger whose logger name
  is usually the FQCN of a handler. However, we often do not have much
  control over the FQCN of a handler, because a Service and its codec
  and handler are decorated heavily. Therefore, the service logger name
  should be determined in more sensible yet configurable way than just
  using FQCN.
- In many cases, a user uses the default VirtualHost only to build a
  Server. The current way of building and setting the default
  VirtualHost is not the most convenient.

Modifications:

- Rename ServiceEntry to ServiceConfig to reflect what it does better
- Add ServiceConfig.loggerName so that a user can specify the preferred
  service logger name
- A service logger name is now determined by the PathMapping of a
  Service by default.
- Add the builder method that configures the default VirtualHost to
  ServerBuilder
- ServerBuilder now builds a Server even if no port was added;
  ServerPort(0, HTTP) will be used by default.
- Miscellaneous:
  - Change the parameter of service/codec/handlerAdded() from Server to
    ServiceConfig, which makes more sense
  - Add PathMapping.prefixPath() which is similar to exactPath()
  - Add the parent getters to ServerConfig, VirtualHost, and
    ServiceConfig
  - Rename ServerConfig.services() to serviceConfigs()
  - Rename (ServerConfig|VirtualHost).findService() to
    findServiceConfig()
  - Replace ServiceMapping and MappedService with PathMappings and
    PathMapped

Result:

- Sensible default service logger names
- Better user experience when building a Server

### CODE REVIEW

**Comment by trustin** [comment]
@anuraaga PTAL

File: src/main/java/com/linecorp/armeria/server/PathMapped.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+     *
+     * @param mappedPath the path translated by {@link PathMappings}
+     * @param value      the value
+     */
+    public static <T> PathMapped<T> of(String mappedPath, T value) {
+        return new PathMapped<>(requireNonNull(mappedPath, "mappedPath"),
+                                requireNonNull(value, "value"));
+    }
+
+    private final String mappedPath;
```
Comment (anuraaga) [inline src/main/java/com/linecorp/armeria/server/PathMapped.java]: @Nullable

File: src/main/java/com/linecorp/armeria/server/PathMapped.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    private final T value;
+
+    private PathMapped(String mappedPath, T value) {
```
Comment (anuraaga) [inline src/main/java/com/linecorp/armeria/server/PathMapped.java]: @Nullable for both parameters

File: src/main/java/com/linecorp/armeria/server/PathMapped.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+ */
+
+package com.linecorp.armeria.server;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A value mapped by {@link PathMappings}.
+ */
+public final class PathMapped<T> {
```
Comment (anuraaga) [inline src/main/java/com/linecorp/armeria/server/PathMapped.java]: Consider calling the class PathMappedValue, I'm finding calls like PathMapped.empty() to read strangely.

File: src/main/java/com/linecorp/armeria/server/ServerBuilder.java [hunk 4]
```diff
@@ -43,8 +48,9 @@
  * sb.port(8080, SessionProtocol.HTTP);
  * // Build and add a virtual host.
  * sb.virtualHost(new VirtualHost("*.foo.com").serviceAt(...).build());
- * // Build and add the default virtual host.
- * sb.defaultVirtualHost(new VirtualHost().serviceAt(...).build());
+ * // Add services to the default virtual host.
+ * sb.serviceAt(...);
```
Comment (anuraaga) [inline src/main/java/com/linecorp/armeria/server/ServerBuilder.java]: These are nice - thanks!

File: src/main/java/com/linecorp/armeria/server/ServerBuilder.java [hunk 5]
```diff
@@ -292,10 +424,18 @@ public Server build() {
             blockingTaskExecutor = defaultBlockingTaskExecutor();
         }
 
+        final List<ServerPort> ports =
```
Comment (anuraaga) [inline src/main/java/com/linecorp/armeria/server/ServerBuilder.java]: Add to port()'s docstring that if not called, a default of 0 (randomly-assigned port) and SessionProtocol.HTTP will be used.

File: src/main/java/com/linecorp/armeria/server/ServiceConfig.java [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+     * Creates a new instance.
+     */
+    public ServiceConfig(VirtualHost virtualHost, PathMapping pathMapping, Service service) {
+        this(virtualHost, pathMapping, service, null);
+    }
+
+    /**
+     * Creates a new instance.
+     */
+    public ServiceConfig(VirtualHost virtualHost, PathMapping pathMapping, Service service,
```
Comment (anuraaga) [inline src/main/java/com/linecorp/armeria/server/ServiceConfig.java]: Can you describe what formats are acceptible for LOGGER_NAME_PATTERN? Perhaps it's something like "anything acceptible as a Java class name is acceptible as a logger name"

File: src/main/java/com/linecorp/armeria/server/ServiceConfig.java [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+
+        final Optional<String> prefixOpt = pathMapping.prefixPath();
+        if (prefixOpt.isPresent()) {
+            return normalizeLoggerName(prefixOpt.get());
+        }
+
+        return "__UNKNOWN__";
+    }
+
+    private static String normalizeLoggerName(String loggerName) {
```
Comment (anuraaga) [inline src/main/java/com/linecorp/armeria/server/ServiceConfig.java]: convertPathToLoggerName? Doesn't seem to really normalize, nor is it called when an explicit logger name is set.

File: src/main/java/com/linecorp/armeria/server/PathMapped.java (hunk 3 above)
Comment (trustin) [inline src/main/java/com/linecorp/armeria/server/PathMapped.java]: Let me choose to use `PathMapped` :-)

File: src/main/java/com/linecorp/armeria/server/ServerBuilder.java (hunk 5 above)
Comment (trustin) [inline src/main/java/com/linecorp/armeria/server/ServerBuilder.java]: Good idea :-)

File: src/main/java/com/linecorp/armeria/server/ServiceConfig.java (hunk 6 above)
Comment (trustin) [inline src/main/java/com/linecorp/armeria/server/ServiceConfig.java]: Added the constraints of logger names and prefixes to ServerBuilder.service()/serviceLoggerPrefix() and VirtualHostBuilder.service()

File: src/main/java/com/linecorp/armeria/server/ServerBuilder.java (hunk 5 above)
Comment (trustin) [inline src/main/java/com/linecorp/armeria/server/ServerBuilder.java]: Added the behavior about when port() was never called to the Javadoc of port() methods.

**Comment by trustin** [comment]
@anuraaga Ready for another round.

File: src/main/java/com/linecorp/armeria/server/ServerBuilder.java [hunk 8]
```diff
@@ -284,6 +285,144 @@ public ServerBuilder maxFrameLength(int maxFrameLength) {
...
+        if (defaultVirtualHost != null) {
+            throw new IllegalStateException("ServerBuilder.defaultVirtualHost() invoked already.");
+        }
+    }
+
+    /**
+     * Sets the default {@link VirtualHost}, which is used when no other {@link VirtualHost}s match the
+     * host name of a client request. e.g. the {@code "Host"} header in HTTP or host name in TLS SNI extension
+     *
+     * @throws IllegalStateException if invoked other default {@link VirtualHost} builder methods already,
```
Comment (anuraaga) [inline src/main/java/com/linecorp/armeria/server/ServerBuilder.java]: if any other default {@link VirtualHost} builder methods have been invoked already

File: src/main/java/com/linecorp/armeria/server/ServiceConfig.java [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+        final Optional<String> exactOpt = pathMapping.exactPath();
+        if (exactOpt.isPresent()) {
+            return convertPathToLoggerName(exactOpt);
+        }
+
+        final Optional<String> prefixOpt = pathMapping.prefixPath();
+        return convertPathToLoggerName(prefixOpt);
+    }
+
+    private static String convertPathToLoggerName(Optional<String> loggerNameOpt) {
```
Comment (anuraaga) [inline src/main/java/com/linecorp/armeria/server/ServiceConfig.java]: servicePathOpt

File: src/main/java/com/linecorp/armeria/server/ServiceConfig.java [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+            return "__ROOT__";
+        }
+        loggerName = loggerName.substring(1); // Strip the first slash.
+        if (loggerName.endsWith("/")) {
+            loggerName = loggerName.substring(0, loggerName.length() - 1); // Strip the last slash.
+        }
+
+        final StringBuilder buf = new StringBuilder(loggerName.length());
+        boolean start = true;
+        for (int i = 0; i < loggerName.length(); i ++) {
```
Comment (anuraaga) [inline src/main/java/com/linecorp/armeria/server/ServiceConfig.java]: No space between i ++

File: src/main/java/com/linecorp/armeria/server/ServiceConfig.java [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+
+    static String validateLoggerName(String value, String propertyName) {
+        requireNonNull(value, propertyName);
+        if (!LOGGER_NAME_PATTERN.matcher(value).matches()) {
+            throw new IllegalArgumentException(propertyName + ": " + value);
+        }
+        return value;
+    }
+
+    private static String defaultLoggerName(PathMapping pathMapping) {
```
Comment (anuraaga) [inline src/main/java/com/linecorp/armeria/server/ServiceConfig.java]: This method should have some unit tests.

File: src/main/java/com/linecorp/armeria/server/ServerBuilder.java (hunk 8 above)
Comment (trustin) [inline src/main/java/com/linecorp/armeria/server/ServerBuilder.java]: Ugh, thanks!

File: src/main/java/com/linecorp/armeria/server/ServiceConfig.java (hunk 11 above)
Comment (trustin) [inline src/main/java/com/linecorp/armeria/server/ServiceConfig.java]: Added `ServiceConfigTest`

**Comment by trustin** [comment]
@anuraaga Ready for another round

File: src/main/java/com/linecorp/armeria/server/ServerBuilder.java [hunk 12]
```diff
@@ -284,6 +285,144 @@ public ServerBuilder maxFrameLength(int maxFrameLength) {
...
+        if (defaultVirtualHost != null) {
+            throw new IllegalStateException("ServerBuilder.defaultVirtualHost() invoked already.");
+        }
+    }
+
+    /**
+     * Sets the default {@link VirtualHost}, which is used when no other {@link VirtualHost}s match the
+     * host name of a client request. e.g. the {@code "Host"} header in HTTP or host name in TLS SNI extension
+     *
+     * @throws IllegalStateException if invoked other default {@link VirtualHost} builder methods have been
```
Comment (anuraaga) [inline src/main/java/com/linecorp/armeria/server/ServerBuilder.java]: "if other default" (invoked left in from previous version)

**Pull request merged**